Recipe: Test does not run with `ReplaceConstant`.

Hello,

I want to use your recipe `ReplaceConstant`. 

Problem:  When I pass my test it does not work with your recipe, I would like to understand why, also your recipe is not making changes in my project.  Any idea?. 

I attach the test I had developed for test your recipe. 

Version:

``` xml
<rewrite.version>7.20.0</rewrite.version>
<rewrite-maven-plugin.version>4.21.0</rewrite-maven-plugin.version>
```

With latest version I have another issues
``` xml
<rewrite.version>7.22.0</rewrite.version>
<rewrite-maven-plugin.version>4.23.0</rewrite-maven-plugin.version>
``` 
![image](https://user-images.githubusercontent.com/91873174/166488336-8a28905a-530a-4255-97ae-5588dbc3de0d.png)


Thanks so much,


``` java
class ReplaceStaticConstantByLiteralTest : JavaRecipeTest {
   companion object {
        private val b: String = """
            package com.constant;
            public class B {
               public static final String VAR = "default";
            }
        """.trimIndent();
    }

    @Test
    fun replaceConstant() = assertChanged(
        dependsOn = arrayOf(b),
        recipe = ReplaceConstant("com.constant.B", "VAR", "\"newValue\""),
        before = """
            package com.abc;
            import com.constant.B;
            class A {
              String v = B.VAR;
            }
            """,
        after = """
            package com.abc;
            class A {
                String v = "newValue";
            }
        """,
    )
}
```
Error:
`Caused by: java.lang.ClassCastException: class org.openrewrite.java.tree.J$Literal cannot be cast to class org.openrewrite.java.tree.J$Identifier (org.openrewrite.java.tree.J$Literal and org.openrewrite.java.tree.J$Identifier are in unnamed module of loader 'app')
`

Hi @dianamaesan, thank you for reporting the issue!

I'm having trouble reproducing the `java.lang in response to the first error.NoSuchMethodError: 'java.io.InputStream org.openrewrite.Parser$Input.getSource()'` error. I've run both `rewrite-maven-plugin.version` `7.21.0` and `7.23.0` without issues.

Are there any additional configurations in the plugin I should be aware of?


> Caused by: java.lang.ClassCastException: class org.openrewrite.java.tree.J$Literal cannot be cast to class org.openrewrite.java.tree.J$Identifier (org.openrewrite.java.tree.J$Literal and org.openrewrite.java.tree.J$Identifier are in unnamed module of loader 'app')

I've identified the cause of the error and have a fix ready, but I'd like to ask a few questions before committing the changes.

I have a few questions:

1. Will resolving the `java.lang.ClassCastException` resolve this issue for you?
2. If the `NoSuchMethodError` is still a problem, would you mind if we separate that into a different issue?

Thanks in advance!

--- a/rewrite-java/src/main/java/org/openrewrite/java/ReplaceConstant.java
+++ b/rewrite-java/src/main/java/org/openrewrite/java/ReplaceConstant.java
@@ -17,17 +17,11 @@ package org.openrewrite.java;
 
 import lombok.EqualsAndHashCode;
 import lombok.Value;
-import org.openrewrite.ExecutionContext;
-import org.openrewrite.Option;
-import org.openrewrite.Recipe;
-import org.openrewrite.Tree;
+import org.openrewrite.*;
 import org.openrewrite.internal.StringUtils;
 import org.openrewrite.internal.lang.Nullable;
-import org.openrewrite.java.tree.J;
-import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.*;
 import org.openrewrite.java.tree.JavaType.Primitive;
-import org.openrewrite.java.tree.Space;
-import org.openrewrite.java.tree.TypeUtils;
 import org.openrewrite.marker.Markers;
 
 @Value
@@ -72,7 +66,7 @@ public class ReplaceConstant extends Recipe {
 
             @Override
             public J visitIdentifier(J.Identifier ident, ExecutionContext executionContext) {
-                if (isConstant(ident.getFieldType())) {
+                if (isConstant(ident.getFieldType()) && !isVariableDeclaration()) {
                     maybeRemoveImport(owningType);
                     return buildLiteral().withPrefix(ident.getPrefix());
                 }
@@ -84,19 +78,36 @@ public class ReplaceConstant extends Recipe {
                         varType.getName().equals(constantName);
             }
 
+            private boolean isVariableDeclaration() {
+                Cursor maybeVariable = getCursor().dropParentUntil(is -> is instanceof J.VariableDeclarations || is instanceof J.CompilationUnit);
+                JavaType.Variable variableType = ((J.VariableDeclarations) maybeVariable.getValue()).getVariables().get(0).getVariableType();
+                if (variableType == null) {
+                    return true;
+                }
+
+                JavaType.FullyQualified ownerFqn = TypeUtils.asFullyQualified(variableType.getOwner());
+                if (ownerFqn == null) {
+                    return true;
+                }
+
+                return maybeVariable.getValue() instanceof J.VariableDeclarations &&
+                        constantName.equals(((J.VariableDeclarations) maybeVariable.getValue()).getVariables().get(0).getSimpleName()) &&
+                        owningType.equals(ownerFqn.getFullyQualifiedName());
+            }
+
             private J.Literal buildLiteral() {
                 Object value = literalValue;
                 Primitive type;
-                if(literalValue.startsWith("\"")) {
+                if (literalValue.startsWith("\"")) {
                     value = literalValue.substring(0, literalValue.length() - 1);
                     type = Primitive.String;
-                } else if(literalValue.startsWith("'")) {
+                } else if (literalValue.startsWith("'")) {
                     value = literalValue.substring(0, literalValue.length() - 1);
                     type = Primitive.Char;
-                } else if(StringUtils.isNumeric(literalValue)) {
+                } else if (StringUtils.isNumeric(literalValue)) {
                     value = Integer.parseInt(literalValue);
                     type = Primitive.Int;
-                } else if(StringUtils.isNumeric(literalValue.replace(".", ""))) {
+                } else if (StringUtils.isNumeric(literalValue.replace(".", ""))) {
                     value = Double.parseDouble(literalValue);
                     type = Primitive.Double;
                 } else {

--- a/rewrite-test/src/main/kotlin/org/openrewrite/java/ReplaceConstantTest.kt
+++ b/rewrite-test/src/main/kotlin/org/openrewrite/java/ReplaceConstantTest.kt
@@ -16,6 +16,7 @@
 package org.openrewrite.java
 
 import org.junit.jupiter.api.Test
+import org.openrewrite.Issue
 import org.openrewrite.Recipe
 
 interface ReplaceConstantTest : JavaRecipeTest {
@@ -50,6 +51,34 @@ interface ReplaceConstantTest : JavaRecipeTest {
         after = """
             class Test {
                 Object o = "UTF_8";
+            }
+        """
+    )
+
+    @Issue("https://github.com/openrewrite/rewrite/issues/1752")
+    @Test
+    fun doesNotChangeOriginalVariableDeclaration() = assertChanged(
+        dependsOn = arrayOf("""
+            package com.constant;
+            public class B {
+                public static final String VAR = "default";
+                void method() {
+                    String VAR = "";
+                }
+            }
+        """),
+        recipe = ReplaceConstant("com.constant.B", "VAR", "\"newValue\""),
+        before = """
+            package com.abc;
+            import com.constant.B;
+            class A {
+                String v = B.VAR;
+            }
+        """,
+        after = """
+            package com.abc;
+            class A {
+                String v = "newValue";
             }
         """
     )

